Add low HP notifications

ernierasta opened the pull request:
Hi everyone!

I am on my crusade to improve Flare communication with player. ;-)

This PR add configuration options which allows to turn on 3 kinds of notifications on low health. Player can also set threshold which determines HP percentage when notifications are triggered.

Notification types:

- display a message,
- play a sound,
- change cursor.
![menu](https://user-images.githubusercontent.com/956018/54818743-405caa00-4c9a-11e9-9d48-55935b7920c1.png)

In game it looks like:
![low_health](https://user-images.githubusercontent.com/956018/54818823-726e0c00-4c9a-11e9-9fd3-4b976983f356.png)
Notice cursor and message.
(you can see one of my hud prototypes ;-) )

It solves (it doesn't beep constantly, but maybe cursor will help?): https://github.com/flareteam/flare-engine/issues/1655

I hope you like it.

nwtour commented:
Hello. 

In my dreams, this function looks in UI like:

![lalala](https://user-images.githubusercontent.com/3074285/54825242-26cd5980-4cbe-11e9-9b07-5949579ca066.jpg)

ernierasta commented:
ohh, interesting idea, definitely doable ... hmm, why not, give a second.

I am not sure about option placement ... I agree that this not only Interface thing ...  what anyone else think? Where settings should go?

dorkster commented:
Please keep the options on the Interface tab. Also, I don't think we need separate thresholds for each warning type. I would rather see two options:

- Health warning type. There are 7 options when counting possible combinations of the three notification types (text, sound, cursor) plus 1 for "none".
- Health warning threshold. Increments of 5% would be appropriate.

ernierasta commented:
@dorkster I was thinking about using combo for types, no problem to do it like that, it would be more elegant in GUI, less in settings file, but this is probably not a problem.
Warning threshold is already 5% increment in range: 5%-50%.

@nwtour your concept is nice, but on second thought, @dorkster may be right. Most of players will probably use sound and maybe additional visual hint on the same threshold. And Justin is a veteran here, so I thing that following his opinions will keep Flare consistent.

nwtour commented:
@ernierasta Ok

ernierasta commented:
I have changed to combo and unified warning type to one "low_hp_warning_type" in settings.txt.
Just read edited comment - will push fixed enums.

ernierasta commented:
done

dorkster commented on src/MenuConfig.cpp:
```diff
@@ -287,6 +291,28 @@ MenuConfig::MenuConfig (bool _is_game_state)
 	minimap_lstb->append(msg->get("Hidden"), "");
 
+	// set up low hp notification type combinantions
+	low_hp_warning_lstb->append(msg->get("Disabled"), "Disables all low health notifications.");
```
To reduce translation load, maybe these should have one tooltip string that describes the option as a whole. For example:

>When the player's health is low, they can be notified through a message, sound, and/or cursor change.

Also, it should be wrapped with `msg->get()` to ensure it gets translated.

**But, I would say hold off on changing anything here. I have some more thoughts on which options should be available.**

dorkster commented:
I've commented on a few small things that need changing, but I'd like to share some more of my thoughts on how these options should work.

- How attached are we to having the option to change the cursor on low HP? It has no effect when using the hardware cursor, and I think message + sound (like with low MP) would be sufficient.
- I would still like to see the sound loop. A "low health" sound would likely be something like a heartbeat or a Zelda-style beeping. Maybe looping would be something determined by the modder. Looping the sound also might make the cursor change less necessary.
- If we get rid of the cursor change, maybe the only option on the Interface tab is the threshold. It would contain the percentages and "Disabled". So when enabled, it would show the message and play the sound. Since some players might think the sound is annoying, I would add a checkbox to the Audio tab to disable the low HP sound.

ernierasta commented:
Thank you for review, I will fix it.

* I would personally say, that cursor option is very good and for me most useful type. Sometimes I play in place, where I can not have sound loud or at all, so this is very good visual hint. And you can see, when you recovered. `Gzdoom` has similar option and I always found it very helpfull. Try to play any of [TakeAShower's mods](https://github.com/Smujb/fallen-kingdom) and You will sometime hear only two beeps (when looping works) and finito. :-D It will not work with hardware cursor as any cursor change in Flare.
* Sound looping is good option, but definitely should be optional and as You said configurable on mod level. You can have something like loud "uaagh!" or more subtle looping sound.
* While this would keep options nicely divided by the topic, it may be a little more confusing for user IMHO. Maybe we could introduce new tab? Something more generic like "Gameplay"? "Automatically equip items" would fit also fit that tab better.

But it is up to you. If you want to have like that I can make a change. I have no problem to maintain cursor change as patch for myself.

dorkster commented on src/MenuConfig.cpp:
```diff
@@ -345,6 +371,9 @@ void MenuConfig::init() {
 	cfg_tabs[INTERFACE_TAB].setOptionWidgets(Platform::Interface::COMBAT_TEXT, combat_text_lb, combat_text_cb, msg->get("Show combat text"));
 	cfg_tabs[INTERFACE_TAB].setOptionWidgets(Platform::Interface::AUTO_EQUIP, auto_equip_lb, auto_equip_cb, msg->get("Automatically equip items"));
 	cfg_tabs[INTERFACE_TAB].setOptionWidgets(Platform::Interface::ENTITY_MARKERS, entity_markers_lb, entity_markers_cb, msg->get("Show hidden entity markers"));
+	cfg_tabs[INTERFACE_TAB].setOptionWidgets(Platform::Interface::LOW_HP_WARNING_TYPE, low_hp_warning_lb, low_hp_warning_lstb, msg->get("Select low health notification types"));
```
This label should be "Low health notification" for consistency with "Low health threshold"

dorkster commented on src/Platform.h:
```diff
@@ -70,6 +70,8 @@ class Platform {
 			ENTITY_MARKERS,
 			SHOW_FPS,
 			DEV_MODE,
+			LOW_HP_WARNING_TYPE,
```
I would like these two to be directly after the `LOOT_TOOLTIPS` option.

dorkster commented:
@ernierasta We'll keep the cursor option then. I was just throwing out some ideas. A "Gameplay" tab is an interesting idea, but we can stick with the Interface tab for now.

I added two more review comments. If you can fix these and the previous review comments, I think this will be good to merge.

ernierasta commented on src/StatBlock.cpp:
```diff
@@ -374,6 +375,10 @@ bool StatBlock::loadSfxStat(FileParser *infile) {
 		// @ATTR sfx_levelup|filename|Filename of sound effect for leveling up.
 		sfx_levelup = infile->val;
 	}
+	else if (infile->key == "sfx_lowhp") {
```
It is in `engine/avatar/"+gfx_base+".txt` to allow consistent sounds for **male** vs **female** characters. It would be no problem with looping beeping, but if you use sounds like loud "aghhh", it would sound silly if you use male voice for female character and vice versa. 

This can be solved by having `sfx_lowhp_male` and `sfx_lowhp_female` + additional logic in code, but I am not sure, that is cleaner.

What do you think, maybe you have better solution?

dorkster replied:
Ah, I wasn't thinking of gender-specific sounds. In that case, your solution is what we want to use.

ernierasta commented:
Thanx for additional comments.

Working on sound looping now, when I will figure it out I will update.

ernierasta commented:
Ok, done.

I had to make few changes to SDLSoundManager. If I had analyzed code well, there was never situation, when looping sound has to stop (only specific sound, not all of them). So problem was, that cleanup was unloading sound after stop. I needed to prevent it. At the end, I decided to add optional parameter to play method, hope that is ok. And I also added stopChannel method.

I need to hold loop true/false somewhere, so I've added sfx_lowhp_loop to StatBlock. I was thinking about sfx_lowhp class with value & loop variables, but chosen simpler solution.

dorkster commented:
Merged! I had to make a few fixes to the sound looping code and the tooltip grammar:

- 984cf0a54121933a37813ec12acb873ab1294fb2
- 3b584f8ec1580d3fba62c6f140ee292126147f54

ernierasta commented:
Thank you, great!
